src/tests/testIpAddress.cc

```diff
@@ -532,20 +532,21 @@ TestIpAddress::testtoUrl_fromSockAddr()
     anIPA.toUrl(buf,MAX_IPSTRLEN);
     CPPUNIT_ASSERT( memcmp("192.168.100.12:80", buf, 17) == 0 );
 
-    /* test output when constructed from in6_addr with IPv6 */
-    struct sockaddr_in6 ip6val;
-
-    ip6val.sin6_addr.s6_addr32[0] = htonl(0xC0A8640C);
-    ip6val.sin6_addr.s6_addr32[1] = htonl(0xFFFFFFFF);
-    ip6val.sin6_addr.s6_addr32[2] = htonl(0xFFFFFFFF);
-    ip6val.sin6_addr.s6_addr32[3] = htonl(0xFFFFFFFF);
-    ip6val.sin6_port = htons(80);
-    ip6val.sin6_family = AF_INET6;
+    /* test output when constructed from sockaddr_storage with IPv6 */
+    struct sockaddr_storage ss;
+    auto *ip6val = reinterpret_cast<struct sockaddr_in6*>(&ss);
```

Comment:
Can we do the opposite -- continue to use `sockaddr_in6 ipv6val` definition and then just cast ipv6 to a sockaddr_storage reference at the time of conversion to Ip::Address? Doing so would simplify and clarify this code. Right now, code further below _appears_ to configure one variable (ipv6val) but use another (ss). If we do not have to use sockaddr_storage as ip6val storage while configuring ipv6val, then let's remove that complication.

```suggestion
    struct sockaddr_in6 ip6val;
```

... and adjust Ip::Address construction accordingly.

Same for other similar test cases, of course.

Reply:
Not without adding more technical debt for the future.
`sockaddr_storage` is defined as being large enough to store the largest `sockaddr_*` type. That includes experimental IPv7+ address types and some edge-cases like interface-scoped IPv6. If we went with `sockaddr_in6` we would lock Squid to only supporting "pure" IPv6.

I would like to get the API as minimal as possible without locking out capabilities like that. But not to go as far as a C library with `sockaddr *` raw-pointers.

Reply:
> Not without adding more technical debt for the future.

I do not see any technical debt. I suspect that my suggestion was misinterpreted.

I now also realize that the second part of my suggestion was wrong -- one must not cast a sockaddr_in6 variable to sockaddr_storage. Only the opposite cast is safe!

> `sockaddr_storage` is defined as being large enough to store the largest `sockaddr_*` type. That includes experimental IPv7+ address types and some edge-cases like interface-scoped IPv6. If we went with `sockaddr_in6` we would lock Squid to only supporting "pure" IPv6.

I see no relationship between the facts mentioned above and my suggestion (that we should (continue to) use sockaddr_in6 type to create a sockaddr_in6 address for the test case). We obviously want to test with Ip::Address with IPv6 addresses. Let's (continue to) use the corresponding sockaddr_in6 type for that test. There is no reason to outlaw that type; we ought to support it at least until the last sockaddr_in6 expression is gone from Squid primary code.

* In this PR, this code should continue to (naturally) configure its IPv6 input using sockaddr_in6.

* In a future PR, when we add support for experimental IPv7 addresses, the test code will (naturally) configure its IPv7 input using sockaddr_in7.

We can also add sockaddr_storage tests, of course. They can (and probably should) reuse the bulk of existing sockaddr_in6 and sockaddr_in test code, memcopying configured sockaddr_in... addresses to sockaddr_storage variables.
 
> I would like to get the API as minimal as possible without locking out capabilities like that.

No API changes or "locking out capabilities" suggested in this change request.